### ISSUE TITLE
Concern about SC 2.1.4 in WCAG 2.1

### ISSUE THREAD

**Issue by spanchang** [w3c/wcag#1008]
Hello @alastc The Understanding doc for the SC describes a situation where the spoken words "Hey Kim" triggers actions within an email app: the Y of "hey" archives the current message and K in "Kim" moves down one conversation.
Explanation for the example of the Understanding doc (Youtube example #2 under Resources on https://www.w3.org/WAI/WCAG21/Understanding/character-key-shortcuts.html) is:
"In Twitter, say "James" and the single character key shortcut J moves down one item, the single character key shortcut m opens the message dialog box, and the "es" is printed in the dialog box".
If this is really happening, is it not the fault of the speech input AT that it parses a word into separate characters and triggers an action based on a phonetic interpretation? I can't understand how "Hey" can be interpreted as "Y" or "Kim" as "K".
If a speech input program (AT) recognizes a shortcut key meant for keyboard action instead of command words, it is something that needs correction / adjustment in the assistive technology. It should not require an action by the content author.
Accidental activation too should not be covered. Surely everyone has experienced calling the wrong person unintentionally or activating  a button in an app by mistake on a mobile device  or perhaps activating a link on a Web page by mistake.
I believe this SC is making developers and designers jump through hoops and increasing the cost of testing and ensuring conformance.
Developers build single character key shortcuts to help certain user groups and now the SC requires them to also incorporate methods by which they can be circumvented. This is not reasonable.
Given that, "Web accessibility depends not only on accessible content but also on accessible Web browsers and other user agents", the onus should be placed on the makers of AT in this instance.
Best wishes and respectfully,
Sailesh

Note: This issue was flagged in https://github.com/w3c/wcag/issues/953. Creating this issue with clearer title.

**Comment by alastc** [comment]
Hi @spanchang,

This issue was raised by expert voice-input users, and from what they said (and my limited experience matches this) a voice-input product has to pass through what is spoken to the browser.

Whilst a product like Dragon will detect the keywords for Dragon, anything else is passed through. It doesn't necessarily know whether the browser is going to use it for entering text or something else.

Therefore if an author creates a custom keyboard-command, they are the only ones who can prevent accidental triggers.

We went through these type of questions [over two years ago](https://github.com/w3c/wcag21/issues/69) and I'm not seeing new information in your comment that would trigger re-opening this as an issue for the group now.

**Comment by spanchang** [comment]
I agree  that a non-key-word is part of "anything else is passed through" by a voice input program like Dragon. 
Still there is no  explanation for how "Hey Kim"  is is passed on to trigger: 
"y" archives the current message. "k" moves down one conversation and "m" mutes a message or thread.
No one has uttered or entered  "Y" or "K" or "M".
So how can a series of  three shortcut commands  be triggered in succession ?
On the page referenced in your response it states one can execute shortcuts with a "Press" prefix or with a custom macro command.
The Gmail program will require  Press Y or Press k etc.
True this was discussed  before 2.1 was published, but  I am sorry I do not see answers to my concerns. 
I use voice input regularly on an iOS (and less frequently on a Pixel) to  dictate short messages or populate a textarea ... and sure if someone accidently  comes in or says something or I wander near a TV some extra text gets captured by the build-in voice input feature. 
The responsibility is on the user and should not be passed on to the content author who has implemented a shortcut that definitely helps some user groups.
I also do not see the point of the third alternative in the SC that the shortcut is active only when the element has focus. One will not need a shortcut ... one can simply hit the enter key. 
The shortcuts are primarily for keyboard use. Perhaps the user agent or voice input feature should have a setting to disregard single character shortcuts that can be toggled by the user. Again, the onus should be on the AT and not the content author. 
I reiterate  that this SC adds to the cost of accessibility testing. 
I request  that the merits of  this SC should be reconsidered. The SC should be deprecated or  revised and  moved to AAA. 

Thanks,
Sailesh

**Comment by alastc** [comment]
> No one has uttered or entered "Y" or "K" or "M".

Because if Dragon doesn't recognise a built-in command, it passes through the stream of text. I.e. If you say "Hey kim" and that isn't a recognised command, it sends those letters through.

If it did not, dictation would not work.

"Press" is a recognised command, so when that is heard it changes the mode.

It is different from using voice input (without voice control) on a device beause then you have selected a text-entry mode and it is only doing dictation. Dragon is trying to control the whole computer, so doing commands and dictation dynamically.

> I also do not see the point of the third alternative in the SC

It was because if you have focused on a select drop-down (for example), it does not send the letters through to the page in general, so does not trigger the issue. Allowing for that is useful for selecting an option as well.

> I request that the merits of this SC should be reconsidered.

We can put that to the group, but the voice-input experts who brought it to the group know a lot more about it than either of us, and they made a good case. I don't think anything has changed.

**Comment by spanchang** [comment]
Alastair,
I was hoping to meet and discuss this in person at CSUN 2020 but unfortunately I am not going.
You explained: "Because if Dragon doesn't recognise a built-in command, it passes through the stream of text. I.e. If you say "Hey kim" and that isn't a recognised command, it sends those letters through".
Yes, I understand a program like Dragon is for computer access and also does voice input. It also recognizes the particular user's voice characteristics etc. set up in the profile and calculates what is said using some logic.
The onus is on the Dragon software as well as the user to  make sure it can reliably distinguish commands from input-text.
My only thought is if an assistive technology does not understand an interaction from  the user, or differentiate between a command and input text  it cannot  arbitrarily execute a command or two based on its best guess. It should simply do nothing and perhaps  convey that the last input / interaction was not understood.
Yes, if the focus were in a textbox or a spot where characters can be entered, the best guess approach may work simply because the user may be able to edit that text. No great harm is done then. 
Again, this should be handled by the user agent / At and it is too burdensome for developers to be addressing this. 
I leave it to you to handle  this as you think best. 
Thanks,

**Comment by alastc** [comment]
Hi Sailesh,

Where you said:
> My only thought is if an assistive technology does not understand an interaction from the user, or differentiate between a command and input text it cannot arbitrarily execute a command or two based on its best guess.

Dragon isn't in a position to determine what the web-content does with the input. Either it recognizes a dragon command, or it passes the text through to the web content.

I'm not even sure if it is aware when a text-box is focused. 

Whether Dragon *could* do that I don't know, but there does not seem to be any likelihood of that aspect changing.

Imagine if a screenreader didn't catch your keystrokes when browsing a page, and you were in gmail and press "Q" to go to the main. Instead it goes to search your contacts because the page implemented that as a keyboard shortcut. If the vendor (and all similar vendors) showed no interest in updating that, wouldn't that be a long term issue that benefits from a success criteria?

-Alastair

**Comment by spanchang** [comment]
Hello Alastair,
It appears in principle it appears you are agreeing this is something the user agent / AT should implement when you say:
"Imagine if a screenreader didn't catch your keystrokes ...".
And my response as before is, the onus should be on "accessible Web browsers and other user agents" in this instance and not content authors. Getting browser makers to include a setting to ignore / not ignore single character key shortcuts should be where the effort should be directed in the same way the browser makers are urged to support to support a new ARIA role /property if this aspect is not what the AT can handle. All users not just users of AT like Dragon can benefit from this. An SC is needed for user agents and not for WCAG. 
Thanks again,
Sailesh

**Comment by alastc** [comment]
Hi Sailesh,

Firstly, I don't know if it is possible for the user agent to implement, my impression is that it is either very difficult or not feasible. For a start, the browsers have to support the keyboard short cuts, it would be the assistive technology that would have to work out when/whether to allow them.

Secondly, if the assistive technology did work this out then the 'a mechanism' exception would be met. 

Therefore, in the unlikely event of that happening, we don't need to change the criteria.

-Alastair

**Comment by fstrr** [comment]
This issue looks like it can be closed. If it needs to be re-opened, please do that and convert it to a Discussion.
